codex/blockexchange/engine/engine.nim

```diff
@@ -280,6 +269,20 @@ proc scheduleTasks(b: BlockExcEngine, blocksDelivery: seq[BlockDelivery]) {.asyn
 
           break # do next peer
 
+proc issueCancellations(b: BlockExcEngine, addrs: seq[BlockAddress]) {.async.} =
+  ## Tells neighboring peers that we're no longer interested in a block.
+  trace "Sending block request cancellations to peers", addrs = addrs.len
+
+  let sends = b.peers.mapIt(
+    (peer: it, request: b.network.request.sendWantCancellations(peer = it.id,
```

Comment:
Also, no reason to split this into several statements, better do something like:

```nim
let
  failed = (await allFinished(
    b.peers.mapIt(b.network.request.sendWantCancellations(
      peer = it.id,
      address = addrs
      ....
  ))))
  .filterIt( it.failed )

if failed.len > 0:
  trace "Couldn't cancel blocks for some peers", peers = failed.len
```

Reply:
Yeah, that's more complicated than I hoped for, but I want to log the addresses of the peers for which the send failed, not just the length cause that doesn't really tell me much. The other option would've been to log it at the `BlockExcNetwork` level but felt off as no op is doing logging there.

Do you know of a shorter way to do it? Plus: way register errors as trc?

Reply:
you can break it up into two steps, with `let finished = await allFinished` and then `if finished.filterIt( it.failed ) > 0: ...`

Reply:
Wait, but how can I recover the Peer ID from the Future? I had to build a tuple list with peer ids and futures, then filter on that list later on to recover the peer ids associated to failed futures so I can print them.

Reply:
don't bother printing the peer ids, we just care about the number of failed peers..s

Reply:
I disagree - if I need to correlate log traces while debugging, then I will want to know which peers failed to receive messages.

Reply:
You can get a potentially very large list of peers, it's unreasonable to print them all. Otherwise, the future's index corresponds to the index in the `b.peers` array.

Reply:
This is not something I expect will be printed often, and the size is proportional to the neighborhood of a peer which I don't see being larger than a hundred or so (probably much much smaller?). Plus you'll hardly ever fail delivery to all peers, so likely will usually print a subset of them. Anyhow I won't fight this. I can replace with len tmw morning.

Reply:
Done. I believe this makes the logs of lower quality, though, and less useful if we need them for debugging (i.e., we'll have to go back and add the peer IDs back to figure out what's actually going on). I also do not share your concern that those messages are going to grow large, as peer neighborhoods are likely to be small, and failed sends to be even smaller and more sparse (except perhaps when there is an issue, at which point you probably want to see it loud and clear).

Reply:
Btw, if you're concerned with trace sizes, we [trace every message send that goes through the block exchange](https://github.com/codex-storage/nim-codex/blob/1eebaa4ce336c3dd00db97c2c2b5614665414dd7/codex/blockexchange/network/network.nim#L97), and log the peer ID. This generates orders of magnitude more noise than my sparse error condition trace.

Reply:
OTOH that may mean that I just need to beef up the info on the trace error condition handler of the block exchange. I suppose that if it contained the actual message type, this would be enough for me to have the info I think we need in the logs.